Adding NUCLEO_WB55RG support

LMESTM opened the pull request:
### Description

Here is the first Pull request to enable STM32WB55 support.
https://www.st.com/en/microcontrollers-microprocessors/stm32wb55rg.html

This port includes BLE support with CORDIO stack. 

### Pull request type
    [ ] Fix
    [ ] Refactor
    [X ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@jeromecoutant @pan- @screamerbg @avilei 

### Release Notes

Tests are ongoing but feedback and review are welcome already.

bcostm committed:
NUCLEO_WB55RG: add SDK files

- Contains files from STM32Cube_FW_WB_V1.0.0

bcostm committed:
NUCLEO_WB55RG: update mbed_rtx.h

bcostm committed:
NUCLEO_WB55RG: update STM common files

- Include RTC ll file from hal as in other families
- STM32WB: update Flash API driver

bcostm committed:
NUCLEO_WB55RG: update targets.json

bcostm committed:
NUCLEO_WB55RG: add WB Cube files reused for BLE transport layer

jeromecoutant committed:
NUCLEO_WB55RG: HAL API updates to get SLEEP, RTC and LPTICKER OK

- astyle OK
- file alignment with other families
- HSE, MSI, HSI clock support
- LPTICKER with RTC and LPTIM tested

LMESTM committed:
NUCLEO_WB55RG: WB Transport Layer Cube files modifications

- Need to force ARM packed redefinition
- Configure LL stack in LL mode only
- Remove warning in shci_tl.c (PLACE_IN_SECTION)

LMESTM committed:
NUCLEO_WB55RG: IAR, ARM and GCC linker files alignment

Align all scatter BLE shared memory declarations.

LMESTM committed:
NUCLEO_WB55RG: add Cordio HCI and Transport Layer driver

The STM32WB Coridio driver includes:
- the Cordio HCI driver handling the reset sequence. During reset sequence
the TX POWER level is set and the BD address is defined if found in OTP
or option bytes. The rest of the sequence is based on the standard CORDIO
HCI driver example.
- The Transport Layer part handles sending and receiving messages to the
WB controller running on cortex-M0 of the STM32WB target. The messages
are shared through shared memory and mailboxes system based on IPCC HW.

LMESTM committed:
NUCLEO_WB55RG: Rework Clock and sleep support

- move hw_conf.h file to targets/TARGET_STM/TARGET_STM32WB directory as
this is used also out of BLE feature.
- create a dedicated hal_deepsleep function as the behavior in WB is a lot
different from other existing STM32 targets
- update clock tree configuration to directly clock the entire tree @ 32MHz
out of HSE. This is needed as we want to let the M0 core running without
any change on M0-side of clocks when M4 enters /exits deep sleep.

LMESTM committed:
STM32WB: Move STM32WB utilies from FEATURE_BLE to targets folder

These files are not BLE specific, but also needed for some clock setting
for instance.

In order to compile an MBED2 application, we need to move the files.

LMESTM committed:
STM32WB: FIX LL RTC warning

LMESTM committed:
STM: fix minor warnings

LMESTM committed:
Add WB support and CUBE FW version in readme.md

LMESTM committed:
STM32WB: Add missing analogin_pinmap

This is required since PR #9449
commit
"Add HAL API for analog in pinmap"

LMESTM committed:
fixup! NUCLEO_WB55RG: add SDK files

LMESTM committed:
STM32WB: Adapt I2C timings

for now based on L4+ cubeMX inputs

LMESTM committed:
STM32WB: Fix ARM link error in mbed2

In case of mbed2, BLE feature is not built.

As there is a MAPPING_TABLE in BLE feature which is not compiled in case
of mbed2, the linker reported the below error

[ERROR] "C:/Data/Workspace/mbed/BUILD/test/NUCLEO_WB55RG/ARM/MBED_2/
.link_script.sct", line 65 (column 6): Error: L6236E:
No section matches selector - no section to be FIRST/LAST.

Solution is to check whether BLE is enabled.

LMESTM committed:
STM32WB: add debug trace group in BLE Transport Layer

LMESTM committed:
STM32WB: Set a random static address during init

0xc0170 commented on features/FEATURE_BLE/targets/TARGET_STM/TARGET_NUCLEO_WB55RG/hw.h:
```diff
@@ -0,0 +1,310 @@
+ * <h2><center>&copy; Copyright (c) 2019 STMicroelectronics.
+ * All rights reserved.</center></h2>
+ *
+ * This software component is licensed by ST under Ultimate Liberty license
```
This shall be  licensed under permissive license as stated in our license guide https://os.mbed.com/docs/mbed-os/v5.11/contributing/index.html#licensing

The drivers are "This software component is licensed by ST under BSD 3-Clause license" . It would be good to have this as well.

LMESTM replied:
@0xc0170 WIll check this and update ! thanks

LMESTM replied:
@0xc0170  Ok. We've now agreed for updating the licence headers - which is now done in the PR.

LMESTM committed:
STM32WB: Add TRNG HW Semaphore

Because TRNG is a shared resource between the 2 STM32WB cores, SW needs
to acquire HW Semaphore before using the resource.

LMESTM committed:
STM32WB: Add FLASH HW Semaphore

Because FLASH is a shared resource between the 2 STM32WB cores, SW needs
to acquire HW Semaphore before using the resource.

0xc0170 commented on targets/TARGET_STM/TARGET_STM32WB/TARGET_STM32WB55xG/device/TOOLCHAIN_ARM_MICRO/stm32wb55xx.sct:
```diff
@@ -0,0 +1,73 @@
+  ; SRAM2 - Shared memory
+  RW_IRAM2a 0x20030000 0x00002800  {  ; RW data
+#if defined(BLE)
+   *(MAPPING_TABLE, +First)
```
is the order important here (thus using FIRST attribute) ? I can see its used in IAR as well not for GCC ARM - I do not see there any preprocessor for this section as here (if defined BLE) ?

I suspect you are getting FIRST/LAST error because MAPPING_TABLE is not defined in your code. If BLE is not present, should at least be defined ?

LMESTM replied:
@0xc0170  The #if defined(BLE) actually does not work: the condition is always considered FALSE. This results in MBED2 compiling ok but has created a regression on mbed-os BLE use cases, That's why target_cordio-tests-* are failed now with ARM TC.  

If I remove this condition, cordio tests are ok, but MBED2 link will fail. 

Since then I tried replacing BLE by FEATURE_BLE or other macros names, but still doesn't work. I can't find a way to define a condition related to BLE or mbed-os vs. MBED2 to solve my issue ... that's why I'd like guidance.

And yes the +First attribute is needed as the mapping table needs to be mapped at this exact place. Without FIRST I understand that ARM linker will not ensure the order. I think other linker ensure the order without a specific attribute.  If you think that  my understanding is wrong, then I would welcome any other proposal to test ...

0xc0170 replied:
The order is specified here http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0474l/Chunk364384667.html

Where is MAPPING_TABLE defined?  I cant find it. In the startup you define only these

```
; Vector Table Mapped to Address 0 at Reset
                AREA    RESET, DATA, READONLY
```

0xc0170 replied:
`FEATURE_BLE` should be defined if this feature is enabled

LMESTM replied:
FEATURE_BLE is defined for all C files (BTW it is also defined in MBED2 application even if the BLE feature is not compiled) but it does not seem to be taken into account for linker script ...

If I look at the files .profile-ld it seems to contain several list of macros: 'flags' list, 'macros' list and 'symbols' list.  

FEATURE_BLE is part of the 'symbols' list.

On the other hand, MBED_BOOT_STACK_SIZE which is used as well in linker script, is part of the 'flags' list, not in the macros, nor in the symbols ... 
```
    "flags": [
        "--keep=os_cb_sections", 
        "--predefine=\"-DMBED_BOOT_STACK_SIZE=4096\"", 
        "--show_full_path"
    ], 
```

So I am wondering how to define a 'flag' that will be defined in mbed-os when BLE feature is compiled and not defined otherwise.

LMESTM replied:
> The order is specified here http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0474l/Chunk364384667.html

Which chapter are you pointing to ?

0xc0170 replied:
Wrong copy paste, Default section placement is the chapter .

Taken from there:

```
By default, the linker places input sections in the following order within an execution region:

By attribute as follows:

    Read-only code.

    Read-only data.

    Read-write code.

    Read-write data.

    Zero-initialized data.

By input section name if they have the same attributes. Names are considered to be case-sensitive and are compared in alphabetical order using the ASCII collation sequence for characters.

By a tie-breaker if they have the same attributes and section names. By default, it is the order that armlink processes the section. You can override this with the FIRST or LAST execution region attribute.
```

LMESTM replied:
hmm interesting - that means that alphabetical order should be enough for taking care of my issue here ...

LMESTM replied:
That works fine - thanks @0xc0170  
Much simpler and straightforward than what I was searching for :-) 
I have updated the PR with the change !

0xc0170 replied:
Awesome ! So after all how it was fixed? To know in the future if we have this problem again :-)

LMESTM replied:
I have added an EDIT in previous comment: 
**EDIT: SOLVED with Martin support:** no need to use the +First attribute as the alphabetical order will ensure that MAPPING_TABLE is placed before MB_MEM1 : tested OK and branch updated.

LMESTM committed:
STM32WB: ARM linker script update

There is no need to add FIRST attribute to MAPPING_TABLE as the default
ordering is alphabetical order.

With this change, we don't have any warning with MBED2 and the sections
are properly ordered anyway in BLE cases.

LMESTM committed:
STM32WB55RG: Add board declaration in tools

In travis build and in SW4STM32 exporter

LMESTM committed:
STM32WB55RG: temporarily remove device_name property in targets.json

Until the CMSIS pack device name is officially deployed.

then we'll the name as can be found in Keil CMSIS pack

       <!-- *************************  Device 'STM32WB55RG'  ***************************** -->
        <device Dname="STM32WB55RGVx">
          <memory id="IROM1"                           start="0x08000000" size="0x001000000" startup="1" default="1" />
          <memory id="IRAM1"                           start="0x20000000" size="0x000040000" init="0"    default="1" />
          <algorithm name="CMSIS/Flash/STM32WB_M4.FLM" start="0x08000000" size="0x001000000"             default="1" />

          <feature type="QFP" n="68"/>
        </device>

LMESTM committed:
STM32WB: Update headers

LMESTM committed:
STM32WB: remove duplicate file

screamerbg reviewed (approved):
Looking good to me

0xc0170 reviewed (changes requested):
Thanks for updating the license 👍 

Looks fine, just new files should have SPDX identifier (like file  targets/TARGET_STM/TARGET_STM32WB/TARGET_STM32WB55xG/device/cmsis.h). 

Sorry for the delay, rc1/rc2 are taking as much time as they could this week.

LMESTM committed:
STM32WB: fix compilation issue with ARMC6

LMESTM committed:
STM32WB: Add SPDX identifier to new files

also update the copyright year when needed

LMESTM committed:
STM32WB: move HCI driver to target specific folder

Need to move stm32wb_HCIDriver.cpp to target specific folder otherwise
it could mess up with other STM targets.

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_STM/TARGET_NUCLEO_WB55RG/stm32wb_HCIDriver.cpp:
```diff
@@ -0,0 +1,902 @@
+                if (hciCoreCb.extResetSeq)
+                {
+                    /* send first extended command */
+                    (*hciCoreCb.extResetSeq)(pMsg, opcode);
```
the extResetSeq needs a command to start it off, please have a look at:
https://github.com/paul-szczepanek-arm/mbed-os/commit/b96c0ac986234457c0c2450ce79a3c9bd1336290

LMESTM replied:
Ok - thanks will do

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_STM/TARGET_NUCLEO_WB55RG/stm32wb_HCIDriver.cpp:
```diff
@@ -0,0 +1,902 @@
+        // Note:
+        // 01 command
+        // type 02 ACL DATA not yet supported
+        // type 03 SCO Voice
```
this seems out of place, also why not enum?

LMESTM replied:
I agree - this comes from the reference Cube software that will most probably evolve over time ... so the enum may soon be introduced and I will use it

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_STM/TARGET_NUCLEO_WB55RG/stm32wb_HCIDriver.cpp:
```diff
@@ -0,0 +1,902 @@
+        // 01 command
+        // type 02 ACL DATA not yet supported
+        // type 03 SCO Voice
+        // type 04 event - sens remontant
```
irrelevant? and in French

LMESTM replied:
yes - I will fix this

paul-szczepanek-arm commented on features/FEATURE_BLE/targets/TARGET_STM/TARGET_NUCLEO_WB55RG/stm32wb_HCIDriver.cpp:
```diff
@@ -0,0 +1,902 @@
+/*
```
General comment because I'm not going to point out each line but code style is inconsistent, I assume copy&paste from existing code with a different style. The braces and indent change and the sections are mixed - this could really do with splitting up into a header and cpp file (separate for driver and transport driver) - make this more maintainable in the long run.

LMESTM replied:
Ok sure. I will try to improve code style consistency. 
Do you have a reference implementation of separate driver and transport driver that can be used as a good example?

paul-szczepanek-arm replied:
https://github.com/ARMmbed/mbed-os/tree/master/features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NORDIC_CORDIO/TARGET_NRF5x

LMESTM replied:
Thanks.
About code style, note that the Porting guide template is maybe not fully consistent with mbed coding style ...  https://github.com/armmbed/mbed-os/blob/master/features/FEATURE_BLE/targets/TARGET_CORDIO/doc/PortingGuide.md

paul-szczepanek-arm replied:
thanks, that's embarrassing, I'll make a note to fix that

LMESTM replied:
that's indeed the code I started from ... ;-)

paul-szczepanek-arm reviewed (changes requested):
you may ignore my style complaints - the only change needed is the reset sequence fix, otherwise it will stall if ext sequence exists

LMESTM committed:
STM32WB: handle extended reset

as suggested during review, apply same sequence as in:
https://github.com/paul-szczepanek-arm/mbed-os/commit/b96c0ac986234457c0c2450ce79a3c9bd1336290

LMESTM committed:
STM32WB: fix some styling in HCI driver

LMESTM committed:
STM32WB: update GCC linker script to match with master

LMESTM committed:
STM32WB: Handle re-init case of transport layer

Issue was seen when running BLE_GAP example from
mbed-os-example-ble.

In STM32WB, the M0 core cannot be reset except if the whole target
is reset. So in case of re-initialization of the BLE stack, the
transport layer should not be initialized again. The HCI reset
command will do the job.

LMESTM committed:
STM32WB: update deep sleep sequence

Review HSE clock initialization to match with latest CUBE firmware.
Also there is no need to set the full clock tree again after deep sleep exit.

With this change we get a stable deep sleep mode (when allowed by CORDIO stack).

LMESTM committed:
STM32WB: disable debug lines when not needed

When doing so, do not disbale GPIO clocks as they may be used by other
drivers !

As a result, debug will be disabled by default, but can be enabled by
either modifying code or selecting MBED debug profile.

LMESTM committed:
STM32WB: Only configure default peripherals in SetSysClock

Typically the RTC clock is configured by RTC driver itself.

RNG on the other hand is shared with M0+ core and it is expected that
M4 turns it on at boot time.

LMESTM committed:
STM32WB: ADC INTERNAL CHANNEL reset after read

Internal channels use is enabling ADC "internal path" which needs
to be disabled after measurement.

Same applied here for WB family as was done for others in #10143.

pan- reviewed (approved):
Validated on the BLE side. Thanks @paul-szczepanek-arm

ashok-rao reviewed (approved):
LGTM from a new target perspective. Thanks @LMESTM
